src/datadog/endpoint_guessing.cpp

```diff
@@ -0,0 +1,179 @@
+  is_str = 1 << 4,
+};
+
+std::string_view to_string(component_type type) noexcept {
+  switch (type) {
+    case component_type::is_int:
+      return "{param:int}";
+    case component_type::is_int_id:
+      return "{param:int_id}";
+    case component_type::is_hex:
+      return "{param:hex}";
+    case component_type::is_hex_id:
+      return "{param:hex_id}";
+    case component_type::is_str:
+      return "{param:str}";
+    default:
+      return "";
+  }
+}
+
+inline uint8_t bool2mask(bool x) noexcept {
+  return static_cast<uint8_t>(-int{x});  // 0 -> 0x00, 1 -> 0xFF
+}
+
+component_type component_replacement(std::string_view path) noexcept {
+  // viable_components is a bitset of the component types not yet excluded
+  std::uint8_t viable_components = 0x1F;  // (is_str << 1) - 1
+  bool found_special_char = false;
+  bool found_digit = false;
+
+  if (path.size() < 2) {
+    viable_components &= ~(component_type::is_int | component_type::is_int_id |
+                           component_type::is_hex | component_type::is_hex_id);
+  } else if (path.size() < 3) {
+    viable_components &= ~(component_type::is_int_id | component_type::is_hex |
+                           component_type::is_hex_id);
+  } else if (path.size() < 6) {
+    viable_components &= ~(component_type::is_hex | component_type::is_hex_id);
+  }
+
+  // is_int does not allow a leading 0
+  if (!path.empty() && path[0] == '0') {
+    viable_components &= ~component_type::is_int;
+  }
+
+  for (std::size_t i = 0; i < path.size(); ++i) {
+    char c = path[i];
+    found_special_char = found_special_char || is_str_special(c);
+    found_digit = found_digit || is_digit(c);
+
+    std::uint8_t digit_mask =
+        bool2mask(is_digit(c)) &
+        (component_type::is_int | component_type::is_int_id |
+         component_type::is_hex | component_type::is_hex_id);
+
+    std::uint8_t hex_alpha_mask =
+        bool2mask(is_hex_alpha(c)) &
+        (component_type::is_hex | component_type::is_hex_id);
+
+    std::uint8_t delimiter_mask =
+        bool2mask(is_delim(c)) &
+        (component_type::is_int_id | component_type::is_hex_id);
+
+    viable_components &=
+        (digit_mask | hex_alpha_mask | delimiter_mask | component_type::is_str);
+  }
+
+  // is_str requires a special char or a size >= 20
+  viable_components &= ~component_type::is_str |
+                       bool2mask(found_special_char || (path.size() >= 20));
+  // hex, and hex_id require a digit
+  viable_components &= ~(component_type::is_hex | component_type::is_hex_id) |
+                       bool2mask(found_digit);
+
+  if (viable_components == 0) {
+    return component_type::none;
+  }
+
+  // c++20: use std::countr_zero
+  std::uint8_t lsb = static_cast<std::uint8_t>(
+      viable_components &
+      static_cast<std::uint8_t>(-static_cast<int8_t>(viable_components)));
+  return static_cast<component_type>(lsb);
+}
+}  // namespace
+
+std::string guess_endpoint(std::string_view orig_path) {
```

Comment:
```suggestion
std::string infer_endpoint(std::string_view orig_path) {
```
To keep consistency with the RFC.